scapy/contrib/isotp/isotp_soft_socket.py

```diff
@@ -220,58 +207,56 @@ def select(sockets, remain=None):
 class TimeoutScheduler:
     """A timeout scheduler which uses a single thread for all timeouts, unlike
     python's own Timer objects which use a thread each."""
-    VERBOSE = False
     GRACE = .1
-    _mutex = Lock()
+    _mutex = RLock()
     _event = Event()
     _thread = None  # type: Optional[Thread]
 
     # use heapq functions on _handles!
     _handles = []  # type: List[TimeoutScheduler.Handle]
 
-    @staticmethod
-    def schedule(timeout, callback):
+    @classmethod
+    def schedule(cls, timeout, callback):
         # type: (float, Callable[[], None]) -> TimeoutScheduler.Handle
         """Schedules the execution of a timeout.
 
         The function `callback` will be called in `timeout` seconds.
 
         Returns a handle that can be used to remove the timeout."""
-        when = TimeoutScheduler._time() + timeout
-        handle = TimeoutScheduler.Handle(when, callback)
-        handles = TimeoutScheduler._handles
+        when = cls._time() + timeout
+        handle = cls.Handle(when, callback)
+        handles = cls._handles
 
-        with TimeoutScheduler._mutex:
+        with cls._mutex:
             # Add the handler to the heap, keeping the invariant
             # Time complexity is O(log n)
             heapq.heappush(handles, handle)
             must_interrupt = (handles[0] == handle)
 
             # Start the scheduling thread if it is not started already
-            if TimeoutScheduler._thread is None:
-                t = Thread(target=TimeoutScheduler._task,
-                           name="TimeoutScheduler._task")
+            if cls._thread is None:
+                t = Thread(target=cls._task, name="TimeoutScheduler._task")
                 must_interrupt = False
-                TimeoutScheduler._thread = t
-                TimeoutScheduler._event.clear()
+                cls._thread = t
+                cls._event.clear()
                 t.start()
 
         if must_interrupt:
             # if the new timeout got in front of the one we are currently
             # waiting on, the current wait operation must be aborted and
             # updated with the new timeout
-            TimeoutScheduler._event.set()
+            cls._event.set()
```

Comment:
After pushing a new Timeout, the main thread should be interrupted, which actually never happened in some rare cases.